File: packages/convex/userPreferences.ts
/* eslint-disable ts/ban-ts-comment */
// @ts-nocheck
import { v } from 'convex/values';
import { mutation, query } from './_generated/server';

/**
 * User Preferences - Learning and Personalization
 * Tracks user behavior and preferences for personalized recommendations
 */

// Preference category types
const preferenceCategories = v.union(
  v.literal('food'),           // Food & Dining
  v.literal('culture'),        // Culture & History
  v.literal('nature'),         // Nature & Outdoors
  v.literal('shopping'),       // Shopping
  v.literal('nightlife'),      // Nightlife & Entertainment
  v.literal('adventure'),      // Adventure & Sports
  v.literal('relaxation'),     // Spa & Relaxation
  v.literal('photography'),    // Photography spots
  v.literal('family'),         // Family-friendly
  v.literal('budget'),         // Budget travel
  v.literal('luxury')          // Luxury travel
);

// Get user preferences
export const getUserPreferences = query({
  args: { userId: v.string() },
  handler: async (ctx, args) => {
    const preferences = await ctx.db
      .query('userPreferences')
      .withIndex('by_user', (q) => q.eq('userId', args.userId))
      .first();

    if (!preferences) {
      return null;
    }

    return preferences;
  },
});

// Get or create user preferences (returns default if none exists)
export const getOrCreatePreferences = query({
  args: { userId: v.string() },
  handler: async (ctx, args) => {
    const existing = await ctx.db
      .query('userPreferences')
      .withIndex('by_user', (q) => q.eq('userId', args.userId))
      .first();

    if (existing) {
      return existing;
    }

    // Return default preferences (not persisted until explicit action)
    return {
      _id: null,
      userId: args.userId,
      categoryScores: {},
      explicitPreferences: [],
      travelStyle: 'balanced',
      budgetLevel: 'moderate',
      pacePreference: 'moderate',
      preferLocalFood: true,
      preferOffBeatPlaces: false,
      accessibilityNeeds: false,
      totalInteractions: 0,
      lastUpdated: Date.now(),
    };
  },
});

// Initialize or update user preferences
export const upsertPreferences = mutation({
  args: {
    userId: v.string(),
    explicitPreferences: v.optional(
      v.array(preferenceCategories)
    ),
    travelStyle: v.optional(
      v.union(
        v.literal('adventurous'),
        v.literal('relaxed'),
        v.literal('cultural'),
        v.literal('balanced')
      )
    ),
    budgetLevel: v.optional(
      v.union(
        v.literal('budget'),
        v.literal('moderate'),
        v.literal('luxury')
      )
    ),
    pacePreference: v.optional(
      v.union(
        v.literal('slow'),      // 1-2 activities per day
        v.literal('moderate'),  // 3-4 activities per day
        v.literal('fast')       // 5+ activities per day
      )
    ),
    preferLocalFood: v.optional(v.boolean()),
    preferOffBeatPlaces: v.optional(v.boolean()),
    accessibilityNeeds: v.optional(v.boolean()),
  },
  handler: async (ctx, args) => {
    const existing = await ctx.db
      .query('userPreferences')
      .withIndex('by_user', (q) => q.eq('userId', args.userId))
      .first();

    const now = Date.now();

    if (existing) {
      // Update existing preferences
      const updateData: Record<string, unknown> = {
        lastUpdated: now,
      };

      if (args.explicitPreferences !== undefined) {
        updateData.explicitPreferences = args.explicitPreferences;
      }
      if (args.travelStyle !== undefined) {
        updateData.travelStyle = args.travelStyle;
      }
      if (args.budgetLevel !== undefined) {
        updateData.budgetLevel = args.budgetLevel;
      }
      if (args.pacePreference !== undefined) {
        updateData.pacePreference = args.pacePreference;
      }
      if (args.preferLocalFood !== undefined) {
        updateData.preferLocalFood = args.preferLocalFood;
      }
      if (args.preferOffBeatPlaces !== undefined) {
        updateData.preferOffBeatPlaces = args.preferOffBeatPlaces;
      }
      if (args.accessibilityNeeds !== undefined) {
        updateData.accessibilityNeeds = args.accessibilityNeeds;
      }

      await ctx.db.patch(existing._id, updateData);
      return existing._id;
    }

    // Create new preferences
    return await ctx.db.insert('userPreferences', {
      userId: args.userId,
      categoryScores: {},
      explicitPreferences: args.explicitPreferences ?? [],
      travelStyle: args.travelStyle ?? 'balanced',
      budgetLevel: args.budgetLevel ?? 'moderate',
      pacePreference: args.pacePreference ?? 'moderate',
      preferLocalFood: args.preferLocalFood ?? true,
      preferOffBeatPlaces: args.preferOffBeatPlaces ?? false,
      accessibilityNeeds: args.accessibilityNeeds ?? false,
      totalInteractions: 0,
      createdAt: now,
      lastUpdated: now,
    });
  },
});

// Record a user behavior event (view, save, like, etc.)
export const recordBehavior = mutation({
  args: {
    userId: v.string(),
    behaviorType: v.union(
      v.literal('view'),        // Viewed content
      v.literal('save'),        // Saved/favorited
      v.literal('unsave'),      // Removed from favorites
      v.literal('copy'),        // Copied itinerary
      v.literal('share'),       // Shared content
      v.literal('like'),        // Liked content
      v.literal('unlike'),      // Removed like
      v.literal('search'),      // Searched for term
      v.literal('poi_click'),   // Clicked on POI
      v.literal('poi_add')      // Added POI to itinerary
    ),
    targetType: v.union(
      v.literal('guide'),
      v.literal('itinerary'),
      v.literal('poi'),
      v.literal('city'),
      v.literal('search')
    ),
    targetId: v.string(),
    categories: v.optional(v.array(preferenceCategories)), // Associated categories
    metadata: v.optional(v.object({
      duration: v.optional(v.number()),  // Time spent viewing (seconds)
      scrollDepth: v.optional(v.number()), // 0-100 percent scrolled
      searchQuery: v.optional(v.string()),
      cityName: v.optional(v.string()),
      poiCategory: v.optional(v.string()),
    })),
  },
  handler: async (ctx, args) => {
    const now = Date.now();

    // Record the behavior event
    const eventId = await ctx.db.insert('userBehaviorEvents', {
      userId: args.userId,
      behaviorType: args.behaviorType,
      targetType: args.targetType,
      targetId: args.targetId,
      categories: args.categories ?? [],
      metadata: args.metadata ?? {},
      createdAt: now,
    });

    // Update category scores based on behavior
    if (args.categories && args.categories.length > 0) {
      const preferences = await ctx.db
        .query('userPreferences')
        .withIndex('by_user', (q) => q.eq('userId', args.userId))
        .first();

      // Behavior type weights
      const weights: Record<string, number> = {
        view: 1,
        save: 5,
        unsave: -3,
        copy: 8,
        share: 6,
        like: 4,
        unlike: -2,
        search: 2,
        poi_click: 2,
        poi_add: 7,
      };

      const weight = weights[args.behaviorType] ?? 1;

      if (preferences) {
        // Update existing scores
        const currentScores = (preferences.categoryScores as Record<string, number>) || {};

        for (const category of args.categories) {
          currentScores[category] = (currentScores[category] || 0) + weight;
        }

        await ctx.db.patch(preferences._id, {
          categoryScores: currentScores,
          totalInteractions: (preferences.totalInteractions ?? 0) + 1,
          lastUpdated: now,
        });
      } else {
        // Create new preferences with initial scores
        const initialScores: Record<string, number> = {};
        for (const category of args.categories) {
          initialScores[category] = weight;
        }

        await ctx.db.insert('userPreferences', {
          userId: args.userId,
          categoryScores: initialScores,
          explicitPreferences: [],
          travelStyle: 'balanced',
          budgetLevel: 'moderate',
          pacePreference: 'moderate',
          preferLocalFood: true,
          preferOffBeatPlaces: false,
          accessibilityNeeds: false,
          totalInteractions: 1,
          createdAt: now,
          lastUpdated: now,
        });
      }
    }

    return eventId;
  },
});

// Get user's top preference categories based on behavior
export const getTopCategories = query({
  args: {
    userId: v.string(),
    limit: v.optional(v.number()),
  },
  handler: async (ctx, args) => {
    const preferences = await ctx.db
      .query('userPreferences')
      .withIndex('by_user', (q) => q.eq('userId', args.userId))
      .first();

    if (!preferences || !preferences.categoryScores) {
      return [];
    }

    const scores = preferences.categoryScores as Record<string, number>;
    const sorted = Object.entries(scores)
      .sort(([, a], [, b]) => b - a)
      .slice(0, args.limit ?? 5);

    return sorted.map(([category, score]) => ({
      category,
      score,
      normalized: Math.min(1, score / 50), // Normalize to 0-1 range
    }));
  },
});

// Get recent behavior events for a user
export const getRecentBehaviors = query({
  args: {
    userId: v.string(),
    limit: v.optional(v.number()),
    behaviorType: v.optional(
      v.union(
        v.literal('view'),
        v.literal('save'),
        v.literal('unsave'),
        v.literal('copy'),
        v.literal('share'),
        v.literal('like'),
        v.literal('unlike'),
        v.literal('search'),
        v.literal('poi_click'),
        v.literal('poi_add')
      )
    ),
  },
  handler: async (ctx, args) => {
    let query = ctx.db
      .query('userBehaviorEvents')
      .withIndex('by_user', (q) => q.eq('userId', args.userId));

    const events = await query.order('desc').take(args.limit ?? 50);

    if (args.behaviorType) {
      return events.filter((e) => e.behaviorType === args.behaviorType);
    }

    return events;
  },
});

// Reset user preferences (clear learned data)
export const resetPreferences = mutation({
  args: { userId: v.string() },
  handler: async (ctx, args) => {
    // Find and update preferences (keep explicit settings, reset learned)
    const preferences = await ctx.db
      .query('userPreferences')
      .withIndex('by_user', (q) => q.eq('userId', args.userId))
      .first();

    if (preferences) {
      await ctx.db.patch(preferences._id, {
        categoryScores: {},
        totalInteractions: 0,
        lastUpdated: Date.now(),
      });
    }

    // Delete all behavior events
    const events = await ctx.db
      .query('userBehaviorEvents')
      .withIndex('by_user', (q) => q.eq('userId', args.userId))
      .collect();

    for (const event of events) {
      await ctx.db.delete(event._id);
    }

    return { success: true, deletedEvents: events.length };
  },
});

// Get personalized recommendations based on preferences
export const getRecommendedCategories = query({
  args: { userId: v.string() },
  handler: async (ctx, args) => {
    const preferences = await ctx.db
      .query('userPreferences')
      .withIndex('by_user', (q) => q.eq('userId', args.userId))
      .first();

    if (!preferences) {
      // Return default recommendations for new users
      return {
        topCategories: ['culture', 'food', 'nature'],
        style: 'balanced',
        isLearned: false,
      };
    }

    const scores = (preferences.categoryScores as Record<string, number>) || {};
    const explicit = preferences.explicitPreferences ?? [];

    // Combine learned and explicit preferences
    const combined: Record<string, number> = { ...scores };
    for (const cat of explicit) {
      combined[cat] = (combined[cat] || 0) + 20; // Boost explicit preferences
    }

    const topCategories = Object.entries(combined)
      .sort(([, a], [, b]) => b - a)
      .slice(0, 5)
      .map(([cat]) => cat);

    return {
      topCategories: topCategories.length > 0 ? topCategories : ['culture', 'food', 'nature'],
      style: preferences.travelStyle,
      budgetLevel: preferences.budgetLevel,
      pacePreference: preferences.pacePreference,
      isLearned: preferences.totalInteractions > 10,
      totalInteractions: preferences.totalInteractions,
    };
  },
});

File: apps/ios/Pathfinding/Pathfinding/Models/UserPreferences.swift
import Foundation
import SwiftUI

// MARK: - Preference Category

/// Categories for travel preferences
enum PreferenceCategory: String, Codable, CaseIterable, Identifiable {
  case food = "food"
  case culture = "culture"
  case nature = "nature"
  case shopping = "shopping"
  case nightlife = "nightlife"
  case adventure = "adventure"
  case relaxation = "relaxation"
  case photography = "photography"
  case family = "family"
  case budget = "budget"
  case luxury = "luxury"

  var id: String { rawValue }

  var displayName: String {
    switch self {
    case .food: return "美食"
    case .culture: return "文化历史"
    case .nature: return "自然风光"
    case .shopping: return "购物"
    case .nightlife: return "夜生活"
    case .adventure: return "冒险运动"
    case .relaxation: return "休闲放松"
    case .photography: return "摄影"
    case .family: return "亲子游"
    case .budget: return "经济实惠"
    case .luxury: return "奢华体验"
    }
  }

  var icon: String {
    switch self {
    case .food: return "fork.knife"
    case .culture: return "building.columns"
    case .nature: return "leaf.fill"
    case .shopping: return "bag.fill"
    case .nightlife: return "moon.stars.fill"
    case .adventure: return "figure.hiking"
    case .relaxation: return "sparkles"
    case .photography: return "camera.fill"
    case .family: return "figure.2.and.child.holdinghands"
    case .budget: return "dollarsign.circle"
    case .luxury: return "crown.fill"
    }
  }

  /// Centralised SwiftUI Color for this category.
  /// Single source of truth – replaces all per-view `categoryColor(for:)` switch statements.
  var color: Color {
    switch self {
    case .food: return .orange
    case .culture: return .purple
    case .nature: return .green
    case .shopping: return .pink
    case .nightlife: return .indigo
    case .adventure: return .red
    case .relaxation: return .cyan
    case .photography: return .yellow
    case .family: return .blue
    case .budget: return .mint
    case .luxury: return .orange // iOS has no "gold"; closest is orange
    }
  }
}

// MARK: - Travel Style

/// Travel style preferences
enum TravelStyle: String, Codable, CaseIterable, Identifiable {
  case adventurous = "adventurous"
  case relaxed = "relaxed"
  case cultural = "cultural"
  case balanced = "balanced"

  var id: String { rawValue }

  var displayName: String {
    switch self {
    case .adventurous: return "探险型"
    case .relaxed: return "休闲型"
    case .cultural: return "文化型"
    case .balanced: return "均衡型"
    }
  }

  var description: String {
    switch self {
    case .adventurous: return "喜欢挑战和冒险，探索未知"
    case .relaxed: return "悠闲慢节奏，享受旅途"
    case .cultural: return "深度体验当地文化历史"
    case .balanced: return "兼顾各方面，灵活安排"
    }
  }

  var icon: String {
    switch self {
    case .adventurous: return "figure.climbing"
    case .relaxed: return "cup.and.saucer.fill"
    case .cultural: return "books.vertical.fill"
    case .balanced: return "scale.3d"
    }
  }
}

// MARK: - Budget Level

/// Budget level preferences
enum BudgetLevel: String, Codable, CaseIterable, Identifiable {
  case budget = "budget"
  case moderate = "moderate"
  case luxury = "luxury"

  var id: String { rawValue }

  var displayName: String {
    switch self {
    case .budget: return "经济型"
    case .moderate: return "舒适型"
    case .luxury: return "奢华型"
    }
  }

  var description: String {
    switch self {
    case .budget: return "精打细算，高性价比"
    case .moderate: return "适度消费，品质与价格平衡"
    case .luxury: return "追求极致体验，不吝开销"
    }
  }

  var icon: String {
    switch self {
    case .budget: return "dollarsign"
    case .moderate: return "dollarsign.circle"
    case .luxury: return "dollarsign.circle.fill"
    }
  }
}

// MARK: - Pace Preference

/// Travel pace preferences
enum PacePreference: String, Codable, CaseIterable, Identifiable {
  case slow = "slow"
  case moderate = "moderate"
  case fast = "fast"

  var id: String { rawValue }

  var displayName: String {
    switch self {
    case .slow: return "慢节奏"
    case .moderate: return "适中"
    case .fast: return "快节奏"
    }
  }

  var description: String {
    switch self {
    case .slow: return "每天 1-2 个景点，深度体验"
    case .moderate: return "每天 3-4 个景点，劳逸结合"
    case .fast: return "每天 5+ 个景点，充分利用时间"
    }
  }

  var icon: String {
    switch self {
    case .slow: return "tortoise.fill"
    case .moderate: return "figure.walk"
    case .fast: return "hare.fill"
    }
  }
}

// MARK: - User Preferences

/// User preferences model
struct UserPreferences: Codable, Identifiable {
  let id: String?
  let userId: String
  var categoryScores: [String: Double]
  var explicitPreferences: [PreferenceCategory]
  var travelStyle: TravelStyle
  var budgetLevel: BudgetLevel
  var pacePreference: PacePreference
  var preferLocalFood: Bool
  var preferOffBeatPlaces: Bool
  var accessibilityNeeds: Bool
  var totalInteractions: Int
  var createdAt: Date?
  var lastUpdated: Date

  enum CodingKeys: String, CodingKey {
    case id = "_id"
    case userId = "user_id"
    case categoryScores = "category_scores"
    case explicitPreferences = "explicit_preferences"
    case travelStyle = "travel_style"
    case budgetLevel = "budget_level"
    case pacePreference = "pace_preference"
    case preferLocalFood = "prefer_local_food"
    case preferOffBeatPlaces = "prefer_off_beat_places"
    case accessibilityNeeds = "accessibility_needs"
    case totalInteractions = "total_interactions"
    case createdAt = "created_at"
    case lastUpdated = "last_updated"
  }

  init(from decoder: Decoder) throws {
    let container = try decoder.container(keyedBy: CodingKeys.self)
    id = try container.decodeIfPresent(String.self, forKey: .id)
    userId = try container.decode(String.self, forKey: .userId)
    categoryScores = try container.decodeIfPresent([String: Double].self, forKey: .categoryScores) ?? [:]
    explicitPreferences = try container.decodeIfPresent([PreferenceCategory].self, forKey: .explicitPreferences) ?? []
    travelStyle = try container.decodeIfPresent(TravelStyle.self, forKey: .travelStyle) ?? .balanced
    budgetLevel = try container.decodeIfPresent(BudgetLevel.self, forKey: .budgetLevel) ?? .moderate
    pacePreference = try container.decodeIfPresent(PacePreference.self, forKey: .pacePreference) ?? .moderate
    preferLocalFood = try container.decodeIfPresent(Bool.self, forKey: .preferLocalFood) ?? true
    preferOffBeatPlaces = try container.decodeIfPresent(Bool.self, forKey: .preferOffBeatPlaces) ?? false
    accessibilityNeeds = try container.decodeIfPresent(Bool.self, forKey: .accessibilityNeeds) ?? false
    totalInteractions = try container.decodeIfPresent(Int.self, forKey: .totalInteractions) ?? 0

    if let timestamp = try container.decodeIfPresent(Double.self, forKey: .createdAt) {
      createdAt = Date(timeIntervalSince1970: timestamp / 1000)
    } else {
      createdAt = nil
    }

    if let timestamp = try container.decodeIfPresent(Double.self, forKey: .lastUpdated) {
      lastUpdated = Date(timeIntervalSince1970: timestamp / 1000)
    } else {
      lastUpdated = Date()
    }
  }

  init(
    id: String? = nil,
    userId: String,
    categoryScores: [String: Double] = [:],
    explicitPreferences: [PreferenceCategory] = [],
    travelStyle: TravelStyle = .balanced,
    budgetLevel: BudgetLevel = .moderate,
    pacePreference: PacePreference = .moderate,
    preferLocalFood: Bool = true,
    preferOffBeatPlaces: Bool = false,
    accessibilityNeeds: Bool = false,
    totalInteractions: Int = 0,
    createdAt: Date? = nil,
    lastUpdated: Date = Date()
  ) {
    self.id = id
    self.userId = userId
    self.categoryScores = categoryScores
    self.explicitPreferences = explicitPreferences
    self.travelStyle = travelStyle
    self.budgetLevel = budgetLevel
    self.pacePreference = pacePreference
    self.preferLocalFood = preferLocalFood
    self.preferOffBeatPlaces = preferOffBeatPlaces
    self.accessibilityNeeds = accessibilityNeeds
    self.totalInteractions = totalInteractions
    self.createdAt = createdAt
    self.lastUpdated = lastUpdated
  }

  /// Get top categories sorted by score
  var topCategories: [PreferenceCategory] {
    let sorted = categoryScores.sorted { $0.value > $1.value }
    return sorted.prefix(5).compactMap { PreferenceCategory(rawValue: $0.key) }
  }

  /// Check if preferences have been learned from behavior
  var isLearned: Bool {
    totalInteractions > 10
  }
}

// MARK: - Category Score

/// Category score with normalized value
struct CategoryScore: Codable, Identifiable {
  let category: String
  let score: Double
  let normalized: Double

  var id: String { category }

  var preferenceCategory: PreferenceCategory? {
    PreferenceCategory(rawValue: category)
  }

  enum CodingKeys: String, CodingKey {
    case category
    case score
    case normalized
  }
}

// MARK: - Recommended Categories

/// Personalized recommendations based on preferences
struct RecommendedCategories: Codable {
  let topCategories: [String]
  let style: TravelStyle
  let budgetLevel: BudgetLevel?
  let pacePreference: PacePreference?
  let isLearned: Bool
  let totalInteractions: Int?

  enum CodingKeys: String, CodingKey {
    case topCategories = "top_categories"
    case style
    case budgetLevel = "budget_level"
    case pacePreference = "pace_preference"
    case isLearned = "is_learned"
    case totalInteractions = "total_interactions"
  }

  var topPreferenceCategories: [PreferenceCategory] {
    topCategories.compactMap { PreferenceCategory(rawValue: $0) }
  }
}

// MARK: - Behavior Event

/// User behavior types for tracking
enum BehaviorType: String, Codable {
  case view = "view"
  case save = "save"
  case unsave = "unsave"
  case copy = "copy"
  case share = "share"
  case like = "like"
  case unlike = "unlike"
  case search = "search"
  case poiClick = "poi_click"
  case poiAdd = "poi_add"
}

/// Target types for behavior tracking
enum BehaviorTargetType: String, Codable {
  case guide = "guide"
  case itinerary = "itinerary"
  case poi = "poi"
  case city = "city"
  case search = "search"
}

/// Behavior event metadata
struct BehaviorMetadata: Codable {
  var duration: Int?
  var scrollDepth: Int?
  var searchQuery: String?
  var cityName: String?
  var poiCategory: String?

  enum CodingKeys: String, CodingKey {
    case duration
    case scrollDepth = "scroll_depth"
    case searchQuery = "search_query"
    case cityName = "city_name"
    case poiCategory = "poi_category"
  }
}

/// User behavior event
struct BehaviorEvent: Codable, Identifiable {
  let id: String
  let userId: String
  let behaviorType: BehaviorType
  let targetType: BehaviorTargetType
  let targetId: String
  let categories: [PreferenceCategory]
  let metadata: BehaviorMetadata?
  let createdAt: Date

  enum CodingKeys: String, CodingKey {
    case id = "_id"
    case userId = "user_id"
    case behaviorType = "behavior_type"
    case targetType = "target_type"
    case targetId = "target_id"
    case categories
    case metadata
    case createdAt = "created_at"
  }

  init(from decoder: Decoder) throws {
    let container = try decoder.container(keyedBy: CodingKeys.self)
    id = try container.decode(String.self, forKey: .id)
    userId = try container.decode(String.self, forKey: .userId)
    behaviorType = try container.decode(BehaviorType.self, forKey: .behaviorType)
    targetType = try container.decode(BehaviorTargetType.self, forKey: .targetType)
    targetId = try container.decode(String.self, forKey: .targetId)
    categories = try container.decodeIfPresent([PreferenceCategory].self, forKey: .categories) ?? []
    metadata = try container.decodeIfPresent(BehaviorMetadata.self, forKey: .metadata)

    if let timestamp = try container.decodeIfPresent(Double.self, forKey: .createdAt) {
      createdAt = Date(timeIntervalSince1970: timestamp / 1000)
    } else {
      createdAt = Date()
    }
  }
}

// MARK: - Update Preferences Request

/// Request to update user preferences
struct UpdatePreferencesRequest: Encodable {
  var explicitPreferences: [PreferenceCategory]?
  var travelStyle: TravelStyle?
  var budgetLevel: BudgetLevel?
  var pacePreference: PacePreference?
  var preferLocalFood: Bool?
  var preferOffBeatPlaces: Bool?
  var accessibilityNeeds: Bool?

  enum CodingKeys: String, CodingKey {
    case explicitPreferences = "explicit_preferences"
    case travelStyle = "travel_style"
    case budgetLevel = "budget_level"
    case pacePreference = "pace_preference"
    case preferLocalFood = "prefer_local_food"
    case preferOffBeatPlaces = "prefer_off_beat_places"
    case accessibilityNeeds = "accessibility_needs"
  }
}

// MARK: - Record Behavior Request

/// Request to record a behavior event
struct RecordBehaviorRequest: Encodable {
  let behaviorType: BehaviorType
  let targetType: BehaviorTargetType
  let targetId: String
  var categories: [PreferenceCategory]?
  var metadata: BehaviorMetadata?

  enum CodingKeys: String, CodingKey {
    case behaviorType = "behavior_type"
    case targetType = "target_type"
    case targetId = "target_id"
    case categories
    case metadata
  }
}

// MARK: - API Response Types

struct PreferencesResponse: Codable {
  let success: Bool
  let data: UserPreferences
}

struct CategoryScoresResponse: Codable {
  let success: Bool
  let data: [CategoryScore]
}

struct RecommendationsResponse: Codable {
  let success: Bool
  let data: RecommendedCategories
}

struct BehaviorsResponse: Codable {
  let success: Bool
  let data: [BehaviorEvent]
}

struct RecordBehaviorResponse: Codable {
  let success: Bool
  let data: RecordBehaviorResult
}

struct RecordBehaviorResult: Codable {
  let eventId: String

  enum CodingKeys: String, CodingKey {
    case eventId = "event_id"
  }
}

struct ResetPreferencesResponse: Codable {
  let success: Bool
  let data: ResetPreferencesResult
}

struct ResetPreferencesResult: Codable {
  let success: Bool
  let deletedEvents: Int

  enum CodingKeys: String, CodingKey {
    case success
    case deletedEvents = "deleted_events"
  }
}
